src/typescript/frontend/src/lib/store/event-store.ts

```diff
@@ -305,8 +310,16 @@ export const createEventStore = (initialState: EventState = defaultState) => {
       // Note that we `unshift` here because we add the latest event to the front of the array.
       // We also update the latest bar if the incoming event is a swap or periodic state event.
       pushEventFromClient: (event) => {
-        if (get().guids.has(event.guid)) return;
-        addToLocalStorage(event);
+        if (get().guids.has(event.guid) && !event.guid.startsWith("Swap")) return;
+        if (
+          event.guid.startsWith("Swap") &&
+          (event as Types.SwapEvent).balanceAsFractionOfCirculatingSupply
+        ) {
+          updateLocalStorage(event);
+          return;
+        } else {
+          addToLocalStorage(event);
+        }
```

Comment:
Trying to understand what exactly is the logic here; why are you not returning if the state store already has the `Swap` event?

Also I wrote some type guards for this to get the compile time types:

```typescript
const event: AnyEmojicoinEvent = ...; // some swap event data
const { isSell } = event; // Compiler/IDE would yell at you for this
if (isSwapEvent(event)) {
  // It's now typed as a Types.SwapEvent event
  // You can now access swap-only fields
  const { isSell } = event;
}
```

Reply:
So this is because now you have two types of Swap events. The ones coming from WebSockets, which are the same as before, and the ones coming from the `inbox_swaps` endpoint, which have the balance field. I do this so that if the event does not have the field, it is replaced by the event with the field present.

Reply:
Ahhh okay, gotcha, that makes sense.